coderd/workspaces.go

```diff
@@ -2379,3 +2477,64 @@ func (api *API) publishWorkspaceAgentLogsUpdate(ctx context.Context, workspaceAg
 		api.Logger.Warn(ctx, "failed to publish workspace agent logs update", slog.F("workspace_agent_id", workspaceAgentID), slog.Error(err))
 	}
 }
+
+func validateWorkspaceACLPerms(ctx context.Context, db database.Store, perms map[string]codersdk.WorkspaceRole, field string) []codersdk.ValidationError {
+	// nolint:gocritic // Validate requires full read access to users and groups
+	ctx = dbauthz.AsSystemRestricted(ctx)
+	var validErrs []codersdk.ValidationError
+	for idStr, role := range perms {
+		if err := validateWorkspaceRole(role); err != nil {
+			validErrs = append(validErrs, codersdk.ValidationError{Field: field, Detail: err.Error()})
+			continue
+		}
+
+		id, err := uuid.Parse(idStr)
+		if err != nil {
+			validErrs = append(validErrs, codersdk.ValidationError{Field: field, Detail: idStr + "is not a valid UUID."})
+			continue
+		}
+
+		switch field {
+		case "user_roles":
+			// TODO(lilac): put this back after Kirby button shenanigans are over
+			// This could get slow if we get a ton of user perm updates.
+			// _, err = db.GetUserByID(ctx, id)
+			// if err != nil {
+			// 	validErrs = append(validErrs, codersdk.ValidationError{Field: field, Detail: fmt.Sprintf("Failed to find resource with ID %q: %v", idStr, err.Error())})
+			// 	continue
+			// }
+		case "group_roles":
+			// This could get slow if we get a ton of group perm updates.
+			_, err = db.GetGroupByID(ctx, id)
+			if err != nil {
+				validErrs = append(validErrs, codersdk.ValidationError{Field: field, Detail: fmt.Sprintf("Failed to find resource with ID %q: %v", idStr, err.Error())})
+				continue
+			}
+		default:
+			validErrs = append(validErrs, codersdk.ValidationError{Field: field, Detail: "invalid field"})
+		}
+	}
+
+	return validErrs
+}
```

Comment:
Let's merge this with the template ACL one. Pass in the func like `validateWorkspaceRole` and `validateTemplateRole`.

Reply:
Addressed in https://github.com/coder/coder/pull/19192